Enable debug information by default in CSI

filipw opened the pull request:
**Customer scenario**

Improves script debugging experience 
1) CSI now displays file path and line number for method stack frames defined in .csx files.
2) It is now possible to attach a debugger to csi.exe and step thru the .csx source.

**Bugs this fixes:**

Fixes https://github.com/dotnet/roslyn/issues/13482

**Workarounds, if any**

None.

**Risk**

Low

**How was the bug found?**

Dogfood, customer reports.

### Current

```
λ csi.exe error.csx
hello world
System.AggregateException: One or more errors occurred. ---> System.Exception: crash!!!
   at Submission#0.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
...
```

### With the PR

```
λ csi.exe error.csx
hello world
System.Exception: crash!!!
   at Submission#0.<<Initialize>>d__0.MoveNext() in X:\Documents\dev\csharp-scripting-demos\error.csx:line 3
--- End of stack trace from previous location where exception was thrown ---
...
```

tmat commented on src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs:
```diff
@@ -109,14 +109,17 @@ private int RunInteractiveCore(ErrorLogger errorLogger)
             }
 
             var scriptPathOpt = sourceFiles.IsEmpty ? null : sourceFiles[0].Path;
-            var scriptOptions = GetScriptOptions(_compiler.Arguments, scriptPathOpt, _compiler.MessageProvider, diagnosticsInfos);
```
>var [](start = 12, length = 3)

Why do you move this below? Note that diagnosticsInfos were populated with errors and reported on the next line.

filipw replied:
I would like to be able to answer that question as well 😄

tmat commented on src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs:
```diff
@@ -109,14 +109,17 @@ private int RunInteractiveCore(ErrorLogger errorLogger)
             }
 
+            // only emit symbols for non-interactive mode, and when encoding is known
+            var emitDebugInformation = !_compiler.Arguments.InteractiveMode && code.Encoding != null;
```
> && code.Encoding != null [](start = 75, length = 25)

Encoding should always be known. When is it not the case?

tmat replied:
Why not emit debug info in interactive mode? The line info would be useful when you #load a script, e.g.

---
In reply to: [103508026](https://github.com/dotnet/roslyn/pull/17379#discussion_r103508026) [](ancestors = 103508026)

filipw replied:
correct `var emitDebugInformation = !_compiler.Arguments.InteractiveMode;` would be enough no need to check encoding.
however to emit debug info in interactive, we would need to use some default encoding, since the input doesn't come from a filestream but from string, so at the moment just switching it on causes interactive to fail with "cannot emit debug when encoding is unknown".

tmat commented:
Re tests: yes, we want to have csi covered by tests:
See e.g. http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Scripting.UnitTests/CommandLineRunnerTests.cs,102

filipw commented:
> Re tests: yes, we want to have csi covered by tests:
> See e.g. http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Scripting.UnitTests/CommandLineRunnerTests.cs,102

thanks 👍 I'll throw in some tests then

filipw commented:
@tmat I'm sorry for the delay - I completely forgot about this PR! 😞 😧

I have added the following changes:
 - replaced `Result.ReturnValue` with `GetAwaiter().GetResult().ReturnValue` in CSI script execution. This ensures that when an unhandled error bubbles up from the script it's displayed to the user as an unwrapped "real" exception rather than `AggregateException`
 - added a test to `CsiTests` that verifies that line numbers are shown to the user
 - fixed the script options according to your comments
 - rebased against the latest master

I left the logic to not enable debug information for interactive mode at the moment. The reason is, at the moment interactive mode has its own logic of [handling exceptions](https://github.com/dotnet/roslyn/blob/master/src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs#L313) which is governed by the formatter [here](https://github.com/dotnet/roslyn/blob/master/src/Scripting/Core/Hosting/ObjectFormatter/CommonObjectFormatter.cs#L57-L98). In order to display the line numbers correctly we'll need to modify things there and I'm not sure we want to touch it at that point? (also I believe there is a separate issue tracking that - https://github.com/dotnet/roslyn/issues/5250)

On a side note, I tested `csi.exe` with this change, and aside from line numbers being reported to the users now, we are now also able to successfully connect with a debugger to a running `csi.exe` process, which, I think, is a very useful feature for scripting users.

tmat commented:
@filipw Cool. I'll take a look.

tmat commented on src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs:
```diff
@@ -108,15 +108,20 @@ private int RunInteractiveCore(ErrorLogger errorLogger)
             }
 
+
             var cancellationToken = new CancellationToken();
```
nit: extra new line

filipw replied:
cleaned up 👍

tmat commented:
:shipit:

tmat commented:
I'm gonna update the description to fit our ask mode template.

filipw commented:
For the record, the behavior - due to the removal of `AggregateException` - is now the following:

```
λ csi.exe error.csx
hello world
System.Exception: crash!!!
   at Submission#0.<<Initialize>>d__0.MoveNext() in X:\Documents\dev\csharp-scripting-demos\error.csx:line 3
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.<RunSubmissionsAsync>d__9`1.MoveNext()
```

tmat commented:
@MeiChin-Tsai for ask mode approval (pending tests).

tmat commented:
test windows_coreclr_test_prtest please

MeiChin-Tsai commented:
approved.

tmat commented:
@ivanbasov @jinujoseph FYI

tmat commented:
@filipw Thanks for contribution!

filipw commented:
thanks @tmat!
